pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,415 @@
+        let mut metadata_count: u32 = 0;
+        while metadata_count < received {
+            /* create packet metadata */
+            packets.push(PvPacket::new());
+            // println!("rx_idx: {}", rx_idx);
+            let rx_desc_ptr: *const xdp_desc = unsafe { xsk_ring_cons__rx_desc(&mut nic.rx, rx_idx + metadata_count) }; // bringing information(packet address, packet length) of received packets through descriptors in RX ring
+
+            /* save metadata */
+            packets[metadata_count as usize].start = DEFAULT_HEADROOM;
+            packets[metadata_count as usize].end = DEFAULT_HEADROOM + unsafe { rx_desc_ptr.as_ref().unwrap().len };
+            packets[metadata_count as usize].size = nic.chunk_size;
+            packets[metadata_count as usize].buffer = unsafe { xsk_umem__get_data(nic.buffer, rx_desc_ptr.as_ref().unwrap().addr).cast::<u8>().sub(DEFAULT_HEADROOM as usize) };
+            packets[metadata_count as usize].private = unsafe { (rx_desc_ptr.as_ref().unwrap().addr - DEFAULT_HEADROOM as u64) as *mut c_void };
+            // println!("packet: {:?}", packets[metadata_count as usize]);
+            // packet_dump(&packets[metadata_count as usize]);
+            metadata_count += 1;
+        }
+
+        unsafe { xsk_ring_cons__release(&mut nic.rx, received); } // notify kernel of using filled slots in RX ring as much as **received
+
+        if metadata_count != received {
+            received = metadata_count;
+        }
+    }
+
+    unsafe {
+        if xsk_ring_prod__needs_wakeup(&mut nic.fq) != 0 {
+            libc::recvfrom(xsk_socket__fd(nic.xsk), std::ptr::null_mut::<c_void>(), 0, libc::MSG_DONTWAIT, std::ptr::null_mut::<libc::sockaddr>(), std::ptr::null_mut::<u32>()) ;
+        }
+    }
+
+    received
+}
+
+pub fn pv_send(nic: &mut PvNic, packets: &mut Vec<PvPacket>, batch_size: u32) -> u32 {
+    /* free packet metadata and UMEM chunks as much as the number of filled slots in cq. */
+    let mut cq_idx: u32 = 0;
+    let filled: u32 = unsafe{ xsk_ring_cons__peek(&mut nic.cq, batch_size, &mut cq_idx) }; // fetch the number of filled slots(the number of packets completely sent) in cq
+
+    if filled > 0 {
+        for _ in 0..filled {
+            unsafe { pv_free_(nic, *xsk_ring_cons__comp_addr(&nic.cq, cq_idx)); } // free UMEM chunks as much as the number of sent packets (same as **filled)
+            cq_idx += 1;
+        }
+        unsafe { xsk_ring_cons__release(&mut nic.cq, filled); } // notify kernel that cq has empty slots with **filled (Dequeue)
+    }
+
+     /* reserve TX ring as much as batch_size before sending packets. */
+    let mut tx_idx: u32 = 0;
+    let reserved: u32 = unsafe { xsk_ring_prod__reserve(&mut nic.tx, batch_size, &mut tx_idx) };
+    // println!("packets len: {}, tx reserved: {}, tx_idx: {}", packets.len(), reserved, tx_idx);
+
+    /* send packets if TX ring has been reserved with **batch_size. if not, don't send packets and free them */
```

Comment:
This comment confuses me. The comment say "Send packets if ..." But the condition after this is "When not ok"

Reply:
The comment is clearly confusing. 
The comment say the description about from the comment line to the end of line in pv_send().
"if not" part in the comment is saying about the lines between 378 and 388, and the rest part in the comment is about below the line 390

Reply:
Then, this would be better explains:
> If reserved TX ring is not sufficient, Don't send them and return 0. Else, sent packets as intended

Reply:
the suggestion comment is reflected